src/Tools/dotnet-monitor/CollectionRules/ActionOptionsDependencyAnalyzer.cs

```diff
@@ -0,0 +1,255 @@
+
+            public string GetActionResultToken() => string.Concat(ActionReferencePrefix,
+                Action.Name, Separator, ResultName, SubstitutionSuffix);
+        }
+
+        public ActionOptionsDependencyAnalyzer(CollectionRuleContext context)
+        {
+            _ruleContext = context ?? throw new ArgumentNullException(nameof(context));
+        }
+
+        public IList<CollectionRuleActionOptions> GetActionDependencies(int actionIndex)
+        {
+            EnsureDependencies();
+
+            if (_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return properties
+                    .SelectMany(p => p.Value.ActionDependencies)
+                    .DistinctBy(a => a.Action.Name, StringComparer.OrdinalIgnoreCase)
+                    .Select(a => a.Action).ToArray();
+            }
+            return Array.Empty<CollectionRuleActionOptions>();
+        }
+
+        public object SubstituteOptionValues(int actionIndex, object settings)
+        {
+            EnsureDependencies();
+            if (!_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return settings;
+            }
+
+            if (settings is ICloneable cloneable)
+            {
+                settings = cloneable.Clone();
+            }
+            else
+            {
+                _ruleContext.Logger.InvalidSettings(settings.GetType().FullName);
+                return settings;
+            }
+
+            foreach (KeyValuePair<string, PropertyDependency> property in properties)
+            {
+                string newValue = property.Value.OriginalValue;
+                foreach(ActionDependency actionDependency in property.Value.ActionDependencies)
+                {
+                    string actionToken = actionDependency.GetActionResultToken();
+                    if (!_ruleContext.ActionResults.TryGetValue(actionDependency.Action.Name, out CollectionRuleActionResult results))
+                    {
+                        _ruleContext.Logger.InvalidResultReference(actionToken);
+                        continue;
+                    }
+                    if (!results.OutputValues.TryGetValue(actionDependency.ResultName, out string result))
+                    {
+                        _ruleContext.Logger.InvalidResultReference(actionToken);
+                        continue;
+                    }
+                    newValue = newValue.Replace(actionToken, result, StringComparison.OrdinalIgnoreCase);
+                }
+                property.Value.Property.SetValue(settings, newValue);
+            }
+
+            return settings;
+        }
+
+        private void EnsureDependencies()
+        {
+            if (_dependencies == null)
+            {
+                _dependencies = new Dictionary<int, Dictionary<string, PropertyDependency>>();
```

Comment:
Add _ruleContext.Options.Actions.Count to the ctor to reduce number of reallocations inside the dictionary.